src/dc_imex.rs

```diff
@@ -805,115 +749,92 @@ unsafe fn export_backup(context: &Context, dir: *const libc::c_char) -> bool {
-    // a pointer inside buf2, MUST NOT be free()'d
-    let mut buf2_headerline: *const libc::c_char = ptr::null_mut();
-    if !dir_name.is_null() {
-        let dir = std::path::Path::new(as_str(dir_name));
-        if let Ok(dir_handle) = std::fs::read_dir(dir) {
-            for entry in dir_handle {
-                if entry.is_err() {
+    if dir_name.is_empty() {
+        return false;
+    }
+    let mut set_default: bool;
+    let mut key: String;
+    let mut imported_cnt = 0;
+
+    let dir = std::path::Path::new(dir_name);
+    if let Ok(dir_handle) = std::fs::read_dir(dir) {
+        for entry in dir_handle {
+            let entry_fn = match entry {
+                Err(err) => {
+                    info!(context, "file-dir error: {}", err);
                     break;
                 }
-                let entry_fn = entry.unwrap().file_name();
-                let name_f = entry_fn.to_string_lossy();
-
-                match dc_get_filesuffix_lc(&name_f) {
-                    Some(suffix) => {
-                        if suffix != ".asc" {
-                            continue;
-                        }
-                    }
-                    None => {
+                Ok(res) => res.file_name(),
+            };
+            let name_f = entry_fn.to_string_lossy();
+            let path_plus_name = dir.join(&entry_fn);
+            match dc_get_filesuffix_lc(&name_f) {
+                Some(suffix) => {
+                    if suffix != "asc" {
                         continue;
                     }
+                    set_default = if name_f.contains("legacy") {
+                        info!(context, "found legacy key '{}'", path_plus_name.display());
+                        false
+                    } else {
+                        true
+                    }
                 }
-                let path_plus_name = dir.join(&entry_fn);
-                info!(context, "Checking: {}", path_plus_name.display());
-
-                free(buf.cast());
-                buf = ptr::null_mut();
-
-                if let Ok(buf_r) = dc_read_file(context, &path_plus_name) {
-                    buf = buf_r.as_ptr() as *mut _;
-                    std::mem::forget(buf_r);
-                } else {
+                None => {
                     continue;
-                };
-
-                private_key = buf;
-
-                free(buf2 as *mut libc::c_void);
-                buf2 = dc_strdup(buf);
-                if dc_split_armored_data(
+                }
+            }
+            if let Ok(content) = dc_read_file(context, &path_plus_name) {
+                key = String::from_utf8_lossy(&content).to_string();
+            } else {
+                continue;
+            };
+
+            /* only import if we have a private key */
+            let mut buf2_headerline = String::default();
+            let split_res: bool;
+            unsafe {
+                let buf2 = "".strdup();
```

Comment:
this is not correct, `buf2` should contain the data from the file we read, not an empty string

Reply:
good catch, fixed.  the logic is a bit weird there FWIW.